README.md
# meitrack-parser

[![npm version](https://img.shields.io/npm/v/meitrack-parser.svg?style=flat-square)](https://www.npmjs.com/package/meitrack-parser)
[![npm downloads](https://img.shields.io/npm/dm/meitrack-parser.svg?style=flat-square)](https://www.npmjs.com/package/meitrack-parser)
[![Build Status](https://img.shields.io/travis/lgaticaq/meitrack-parser.svg?style=flat-square)](https://travis-ci.org/lgaticaq/meitrack-parser)
[![Coverage Status](https://img.shields.io/coveralls/lgaticaq/meitrack-parser/master.svg?style=flat-square)](https://coveralls.io/github/lgaticaq/meitrack-parser?branch=master)
[![Code Climate](https://img.shields.io/codeclimate/github/lgaticaq/meitrack-parser.svg?style=flat-square)](https://codeclimate.com/github/lgaticaq/meitrack-parser)
[![dependency Status](https://img.shields.io/david/lgaticaq/meitrack-parser.svg?style=flat-square)](https://david-dm.org/lgaticaq/meitrack-parser#info=dependencies)
[![devDependency Status](https://img.shields.io/david/dev/lgaticaq/meitrack-parser.svg?style=flat-square)](https://david-dm.org/lgaticaq/meitrack-parser#info=devDependencies)

> Parse raw data from Meitrack devices

## Installation

```bash
npm i -S meitrack-parser
```

## Use

[Try on Tonic](https://tonicdev.com/npm/meitrack-parser)
```js
import meitrack from 'meitrack-parser'

const raw = new Buffer('$$A138,862170013556541,AAA,35,7.092076,79.960473,140412132808,A,10,9,57,275,1,14,5783799,7403612,413|1|F6E0|3933,0000,000B|0009||02D8|0122,*EE\r\n');
const data = meitrack.parse(raw);
/*
{ 
  raw: '$$A138,862170013556541,AAA,35,7.092076,79.960473,140412132808,A,10,9,57,275,1,14,5783799,7403612,413|1|F6E0|3933,0000,000B|0009||02D8|0122,*EE\r\n',
  type: 'data',
  manufacturer: 'meitrack',
  device: 'MVT380',
  imei: 862170013556541,
  command: 'AAA',
  alarm: { type: 'Gps' },
  loc: { type: 'Point', coordinates: [ 79.960473, 7.092076 ] },
  datetime: Sat Apr 12 2014 11:28:08 GMT-0200 (CLST),
  gpsStatus: true,
  satellites: 10,
  gsmSignal: 9,
  speed: 57,
  direction: 275,
  hdop: 1,
  altitude: 14,
  odometer: 5783799,
  runtime: 7403612,
  mcc: '413',
  mnc: '1',
  lac: 63200,
  cid: 14643,
  status: { 
    input: {
      '1': false,
      '2': false,
      '3': false,
      '4': false,
      '5': false,
      '6': false,
      '7': false,
      '8': false
    },
    output: {
      '1': false,
      '2': false,
      '3': false,
      '4': false,
      '5': false,
      '6': false,
      '7': false,
      '8': false
    }
  },
  voltage: {
    ad1: 0.064453125,
    ad2: 0.052734375,
    ad3: null,
    battery: 4.265625,
    inputCharge: 13.59375
  }
}
*/
```

## License

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
## 2.0.0 (2017-11-21)

##### Chores

* **package:**
  * update dependencies ([db43c049](https://github.com/lgaticaq/meitrack-parser/commit/db43c0498629a30f0d5a2d276bec6db7d8c50993))
  * update dependencies ([2d1bea04](https://github.com/lgaticaq/meitrack-parser/commit/2d1bea04a3457c124e3a3c1c9cec061d00bafdd4))
  * remove coveralls and codeclimate ([60ce2b1f](https://github.com/lgaticaq/meitrack-parser/commit/60ce2b1f1782c633a0a3fb6182e0c12f95fe7eba))
  * remove moment ([1e4c3c05](https://github.com/lgaticaq/meitrack-parser/commit/1e4c3c05602d2f6663e700e822401ee393533fcc))
  * update eslint to version 4.0.0 ([95e6387a](https://github.com/lgaticaq/meitrack-parser/commit/95e6387aa74e655a9db056b8e2170e070dcaf634))
  * update chai to version 4.0.0 ([244c7d83](https://github.com/lgaticaq/meitrack-parser/commit/244c7d83bcedbd325252542b7edb680698675bb1))
  * update eslint to version 3.9.1 ([8c580dd5](https://github.com/lgaticaq/meitrack-parser/commit/8c580dd5125fe7e0517e7e5b5fb165ba8257e9bb))
  * update crc to version 3.4.1 ([921a3062](https://github.com/lgaticaq/meitrack-parser/commit/921a30623569971748e5fdcee1bb67b5647b27bd))
  * update eslint to version 3.8.0 ([23666a1f](https://github.com/lgaticaq/meitrack-parser/commit/23666a1f1ceafa9ba36156ed72f97c73292537bf))
* **deps:**
  * update dependency eslint to v4.11.0 ([9f344779](https://github.com/lgaticaq/meitrack-parser/commit/9f3447799efcc580eb842feaa497d7fa97c246d6))
  * update dependency eslint to v4.9.0 ([71343a51](https://github.com/lgaticaq/meitrack-parser/commit/71343a510502663b2f3b25133e93e7e19d5c0aa2))
  * update dependency mocha to v4.0.1 ([f25957ba](https://github.com/lgaticaq/meitrack-parser/commit/f25957ba952f55e30ee07045181a7ce0921cec62))
  * update dependency mocha to v4.0.0 ([36b38849](https://github.com/lgaticaq/meitrack-parser/commit/36b388496b9883e88d753b12760b19466d0671e3))
  * update dependency eslint to v4.8.0 ([4148820b](https://github.com/lgaticaq/meitrack-parser/commit/4148820b94a2f0f072c4cc76e7176ef7ab9b584d))
  * update dependency coveralls to v3.0.0 ([52f7fe74](https://github.com/lgaticaq/meitrack-parser/commit/52f7fe7423dca940feb4e30a68900e8ce81ffdfe))
  * update dependency coveralls to v2.13.3 ([2db61c17](https://github.com/lgaticaq/meitrack-parser/commit/2db61c17f4a0f62acc5a4dee245dbdc256fa10bb))
  * update dependency coveralls to v2.13.2 ([5773ea1a](https://github.com/lgaticaq/meitrack-parser/commit/5773ea1aa830f6ca4120733855af823c2386bf7d))
  * update dependency generate-changelog to v1.5.0 ([0cb2f1d4](https://github.com/lgaticaq/meitrack-parser/commit/0cb2f1d4bd15c56b247f99fa02043e839cdc406f))
  * update dependency eslint to v4.7.2 ([40578660](https://github.com/lgaticaq/meitrack-parser/commit/4057866006c6ab15d2607ae9afc2a254d4bc686b))
  * update dependency eslint to v4.7.1 ([9d1132da](https://github.com/lgaticaq/meitrack-parser/commit/9d1132dad2d2b04c792ab453f2aeeb6de622c48f))
  * update dependency eslint to v4.7.0 ([04b6734f](https://github.com/lgaticaq/meitrack-parser/commit/04b6734f2a19223fe25760e4c38dc0858660d395))
  * update dependency mocha to v3.5.3 ([bbf967a0](https://github.com/lgaticaq/meitrack-parser/commit/bbf967a021a09c504fc7eb8063275c0619a00aea))
  * update dependency mocha to v3.5.2 ([29ad74e7](https://github.com/lgaticaq/meitrack-parser/commit/29ad74e74a5fb329877fc32cc04df73188a6aad5))
  * update dependency mocha to v3.5.1 ([f2745412](https://github.com/lgaticaq/meitrack-parser/commit/f27454128b4306d89fb40c562bc8ff2af1a34488))
  * update dependency codeclimate-test-reporter to v0.5.0 ([fdfb2ed4](https://github.com/lgaticaq/meitrack-parser/commit/fdfb2ed48acbfb3050dcfb89fed13183497e2210))

##### New Features

* **node:** update min node version to v6 ([051f2028](https://github.com/lgaticaq/meitrack-parser/commit/051f202845d121ac3a04d35886df092f3b958655))

##### Bug Fixes

* **deps:** update dependency pad to v^2.0.0 ([a32b1bbc](https://github.com/lgaticaq/meitrack-parser/commit/a32b1bbc58d9c17a641802e0b2cb419e36c76957))

##### Refactors

* **deps:** renovate pin dependencies packages ([7042282e](https://github.com/lgaticaq/meitrack-parser/commit/7042282ea5ea05473942e5861663a4ba9bf84c2d))

##### Code Style Changes

* **standard:** migrate to standard ([663cd198](https://github.com/lgaticaq/meitrack-parser/commit/663cd198a250e7a413929a0c96df2c5ec89b012a))

## 1.0.0 (2016-10-20)

##### Chores

* **package:**
  * Update eslint to 3.8.1 ([34a28071](https://github.com/lgaticaq/meitrack-parser/commit/34a280719b60a5f13b3b691746d721c51a88dddc))
  * update codeclimate-test-reporter to version 0.4.0 ([c717f081](https://github.com/lgaticaq/meitrack-parser/commit/c717f0811545e90821b056ef2e1f1fff87d656eb))
  * update eslint to version 3.7.0 ([8d7f80bf](https://github.com/lgaticaq/meitrack-parser/commit/8d7f80bf58d050a48dfaca1ac567e3f5c3693173))
  * update eslint to version 3.6.1 ([f41ffc21](https://github.com/lgaticaq/meitrack-parser/commit/f41ffc216c18bfc631992ff72642a1a7cd66e9e4))

##### Bug Fixes

* **parser:**
  * Remove ci and add cid ([c76303f4](https://github.com/lgaticaq/meitrack-parser/commit/c76303f40c0e3a34ae6cdbce0018e7bcd173a95a))
  * Remove gpsSignal and add gpsStatus ([e419dd68](https://github.com/lgaticaq/meitrack-parser/commit/e419dd681c21b671f9cad1ac2cd694065cdff826))

#### 0.4.1 (2016-10-14)

##### Chores

* **package:**
  * Update dependencies ([08cc569e](https://github.com/lgaticaq/meitrack-parser/commit/08cc569e5eb2708ab00c2bb502f1e9ff21ab7bd0))
  * Add support to yarnpkg ([88e2692a](https://github.com/lgaticaq/meitrack-parser/commit/88e2692a62460b49aac9bd2e578f7324e02b1a79))

##### New Features

* **parser:** Add manufacturer to data ([3343a9af](https://github.com/lgaticaq/meitrack-parser/commit/3343a9af3b9d5ccfc28e7499b3325420e95d52ba))

### 0.4.0 (2016-8-30)

##### Chores

* **package:**
  * Update dependencies ([1e978976](https://github.com/lgaticaq/meitrack-parser/commit/1e9789767e181fd0db9ace957a6fe45af7306090))

test/test.js
'use strict'

const meitrack = require('../src')
const { expect } = require('chai')
const { describe, it } = require('mocha')

describe('meitrack-parser', () => {
  it('should return the raw data parsed', () => {
    const raw = Buffer.from(
      '$$e155,867965021508656,AAA,35,-33.361133,-70.514245,160412155005,A,8,27,0,289,1.3,867,318379,2338885,730|1|32D3|A03F,0008,0000|0000|0000|02DA|0106,00000001,*3A\r\n'
    )
    const data = meitrack.parse(raw)
    expect(data.raw).to.eql(raw.toString())
    expect(data.manufacturer).to.eql('meitrack')
    expect(data.device).to.eql('MVT380')
    expect(data.type).to.eql('data')
    expect(data.imei).to.eql(867965021508656)
    expect(data.command).to.eql('AAA')
    expect(data.alarm.type).to.eql('Gps')
    expect(data.loc.type).to.eql('Point')
    expect(data.loc.coordinates).to.eql([-70.514245, -33.361133])
    expect(data.datetime).to.eql(new Date('2016-04-12T15:50:05.000Z'))
    expect(data.gpsStatus).to.eql(true)
    expect(data.satellites).to.eql(8)
    expect(data.gsmSignal).to.eql(27)
    expect(data.speed).to.eql(0)
    expect(data.direction).to.eql(289)
    expect(data.hdop).to.eql(1.3)
    expect(data.altitude).to.eql(867)
    expect(data.odometer).to.eql(318379)
    expect(data.runtime).to.eql(2338885)
    expect(data.mcc).to.eql('730')
    expect(data.mnc).to.eql('1')
    expect(data.lac).to.eql(13011)
    expect(data.cid).to.eql(41023)
    expect(data.status.input['1']).to.eql(false)
    expect(data.status.input['2']).to.eql(false)
    expect(data.status.input['3']).to.eql(false)
    expect(data.status.input['4']).to.eql(false)
    expect(data.status.input['5']).to.eql(false)
    expect(data.status.input['6']).to.eql(false)
    expect(data.status.input['7']).to.eql(false)
    expect(data.status.input['8']).to.eql(false)
    expect(data.status.output['1']).to.eql(false)
    expect(data.status.output['2']).to.eql(false)
    expect(data.status.output['3']).to.eql(false)
    expect(data.status.output['4']).to.eql(true)
    expect(data.status.output['5']).to.eql(false)
    expect(data.status.output['6']).to.eql(false)
    expect(data.status.output['7']).to.eql(false)
    expect(data.status.output['8']).to.eql(false)
  })

  it('should return the true in data', () => {
    const raw = Buffer.from(
      '$$A138,862170013556541,AAA,35,7.092076,79.960473,140412132808,A,10,9,57,275,1,14,5783799,7403612,413|1|F6E0|3933,0000,000B|0009||02D8|0122,*EE\r\n'
    )
    const data = meitrack.isMeitrack(raw)
    expect(data).to.eql(true)
  })

  it('should return the true in response', () => {
    const raw = Buffer.from('$$y28,867965021508656,C01,OK*31\r\n')
    const data = meitrack.isMeitrack(raw)
    expect(data).to.eql(true)
  })

  it('should return ok data', () => {
    const raw = Buffer.from('$$j28,353358017784062,F01,OK*19\r\n')
    const data = meitrack.parse(raw)
    expect(data.manufacturer).to.eql('meitrack')
    expect(data.device).to.eql('MEITRACK-COMMAND-OK')
    expect(data.type).to.eql('ok')
    expect(data.code).to.eql('F01')
  })

  it('should return command data', () => {
    const command = {
      instruction: '1_on',
      imei: 353358017784062
    }
    const data = meitrack.parseCommand(command)
    expect(data).to.match(
      /^@@([\x41-\x7A])(\d{1,3}),353358017784062,C01,0,12222\*([0-9A-F]{2})\r\n$/
    )
  })

  it('should return command reboot', () => {
    const data = meitrack.getRebootCommand(353358017784062)
    expect(data).to.match(
      /^@@([\x41-\x7A])(\d{1,3}),353358017784062,F02\*([0-9A-F]{2})\r\n$/
    )
  })

  it('should return null imei', () => {
    const raw = Buffer.from('askdhaskjdhakjdhaksjdhaksjdh')
    const imei = meitrack.getImei(raw)
    expect(imei).to.eql(null)
  })

  it('should return MVT380 imei', () => {
    const raw = Buffer.from(
      '$$e155,867965021508656,AAA,35,-33.361133,-70.514245,160412155005,A,8,27,0,289,1.3,867,318379,2338885,730|1|32D3|A03F,0008,0000|0000|0000|02DA|0106,00000001,*3A\r\n'
    )
    const imei = meitrack.getImei(raw)
    expect(imei).to.eql('867965021508656')
  })
})


src/index.js
'use strict'

const crc = require('crc')
const pad = require('pad')

const patterns = {
  mvt380: /^\$\$([\x41-\x7A])(\d{1,3}),(\d{15}),([0-9A-F]{3}),(\d{1,3}),([-]?\d+\.\d+),([-]?\d+\.\d+),(\d{12}),([AV]),(\d{1,3}),(\d{1,2}),(\d+(\.\d+)?),(\d+(\.\d+)?),(\d+(\.\d+)?),(\d+(\.\d+)?),(\d+(\.\d+)?),(\d+),(\d{3})\|(\d{1,3})\|([0-9A-F]{4})\|([0-9A-F]{4}),([0-9A-F]{4}),([0-9A-F]{1,4})?\|([0-9A-F]{1,4})?\|([0-9A-F]{1,4})?\|([0-9A-F]{1,4})\|([0-9A-F]{1,4}),([0-9A-F]{8})?,?([0-9A-F]+)?,?(\d{1,2})?,?([0-9A-F]{4})?,?([0-9A-F]{6})?\|?([0-9A-F]{6})?\|?([0-9A-F]{6})?\|?\*([0-9A-F]{2})\r\n$/,
  ok: /^\$\$([\x41-\x7A])(\d{1,3}),(\d{15}),([0-9A-F]{3}),OK\*([0-9A-F]{2})\r\n$/
}

const parseAlrm = event => {
  const alarms = {
    '1': { type: 'SOS_Button' },
    '2': { type: 'DI', number: 2, status: true },
    '3': { type: 'DI', number: 3, status: true },
    '4': { type: 'DI', number: 4, status: true },
    '5': { type: 'DI', number: 5, status: true },
    '9': { type: 'DI', number: 1, status: false },
    '10': { type: 'DI', number: 2, status: false },
    '11': { type: 'DI', number: 3, status: false },
    '12': { type: 'DI', number: 4, status: false },
    '13': { type: 'DI', number: 5, status: false },
    '17': { type: 'DI', number: 2, status: false },
    '18': { type: 'lowExternalBattery' },
    '19': { type: 'Over_Speed', status: true },
    '20': { type: 'Geo_Fence', status: true },
    '21': { type: 'Geo_Fence', status: false },
    '22': { type: 'Charge', status: true },
    '23': { type: 'Charge', status: false },
    '24': { type: 'gpsSignal', status: false },
    '25': { type: 'gpsSignal', status: true },
    '26': { type: 'Sleep', status: true },
    '27': { type: 'Sleep', status: false },
    '28': { type: 'gpsAntennaCut' },
    '29': { type: 'deviceReboot' },
    '31': { type: 'Heartbeat' },
    '32': { type: 'Angle' },
    '33': { type: 'distanceIntervalTracking' },
    '34': { type: 'replyCurrent' },
    '35': { type: 'Gps' },
    '36': { type: 'tow' },
    '37': { type: 'Rfid' },
    '39': { type: 'picture' },
    '40': { type: 'powerOff' },
    '41': { type: 'Moving', status: false },
    '42': { type: 'Moving', status: true },
    '44': { type: 'jamming', status: true },
    '50': { type: 'temperature', status: true },
    '51': { type: 'temperature', status: false },
    '52': { type: 'fuelFulled' },
    '53': { type: 'fuelEmpty' },
    '54': { type: 'fuelStolen' },
    '56': { type: 'armed' },
    '57': { type: 'disarmed' },
    '58': { type: 'stealing' },
    '63': { type: 'jamming', status: false },
    '65': { type: 'pressInput1ToCall' },
    '66': { type: 'pressInput2ToCall' },
    '67': { type: 'pressInput3ToCall' },
    '68': { type: 'pressInput4ToCall' },
    '69': { type: 'pressInput5ToCall' },
    '70': { type: 'rejectIncomingCall' },
    '71': { type: 'getLocationByCall' },
    '72': { type: 'autoAnswerIncomingCall' },
    '73': { type: 'listenIn' },
    '79': { type: 'fall' },
    '80': { type: 'install' },
    '81': { type: 'dropOff' },
    '139': { type: 'maintenance' }
  }
  return event in alarms ? alarms[event] : null
}

const getMvt380 = raw => {
  const match = patterns.mvt380.exec(raw)
  const status = match[27]
    .split('')
    .map(x => pad(4, parseInt(x, 10).toString(2), '0'))
    .join('')
  const data = {
    raw: match[0],
    type: 'data',
    manufacturer: 'meitrack',
    device: 'MVT380',
    imei: parseInt(match[3], 10),
    command: match[4],
    alarm: parseAlrm(match[5]),
    loc: {
      type: 'Point',
      coordinates: [parseFloat(match[7]), parseFloat(match[6])]
    },
    datetime: new Date(
      match[8].replace(
        /(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})/,
        '20$1-$2-$3T$4:$5:$6.000Z'
      )
    ),
    gpsStatus: match[9] === 'A',
    satellites: parseInt(match[10], 10),
    gsmSignal: parseInt(match[11], 10),
    speed: parseFloat(match[12]),
    direction: parseFloat(match[14]),
    hdop: parseFloat(match[16]),
    altitude: parseFloat(match[18]),
    odometer: parseFloat(match[20]),
    runtime: parseInt(match[22], 10),
    mcc: match[23],
    mnc: match[24],
    lac: parseInt(match[25], 16),
    cid: parseInt(match[26], 16),
    status: {
      output: {
        '1': status[15] === '1',
        '2': status[14] === '1',
        '3': status[13] === '1',
        '4': status[12] === '1',
        '5': status[11] === '1',
        '6': status[10] === '1',
        '7': status[9] === '1',
        '8': status[8] === '1'
      },
      input: {
        '1': status[7] === '1',
        '2': status[6] === '1',
        '3': status[5] === '1',
        '4': status[4] === '1',
        '5': status[3] === '1',
        '6': status[2] === '1',
        '7': status[1] === '1',
        '8': status[0] === '1'
      }
    },
    voltage: {
      ad1: match[28] ? parseInt(match[28], 16) * 6 / 1024 : null,
      ad2: match[29] ? parseInt(match[29], 16) * 6 / 1024 : null,
      ad3: match[30] ? parseInt(match[30], 16) * 6 / 1024 : null,
      battery: match[31] ? parseInt(match[31], 16) * 3 * 2 / 1024 : null,
      inputCharge: match[32] ? parseInt(match[32], 16) * 3 * 16 / 1024 : null
    }
  }
  return data
}

const parseCode = raw => {
  const match = patterns.ok.exec(raw)
  const code = match[4]
  const codes = {
    C01: 'SETIOSWITCH',
    F09: 'CLEARBUF',
    A12: 'SETGPRSINTERVAL',
    B07: 'SETOVERSPEEDALARM',
    F01: 'RBOOT',
    F02: 'RBOOT'
  }
  const data = {
    manufacturer: 'meitrack',
    device: 'MEITRACK-COMMAND-OK',
    type: 'ok',
    code: code
  }
  if (Object.keys(codes).indexOf(code) > -1) data.command = codes[code]
  return data
}

const parse = raw => {
  let result = { type: 'UNKNOWN', raw: raw.toString() }
  if (patterns.mvt380.test(raw)) {
    result = getMvt380(raw)
  } else if (patterns.ok.test(raw)) {
    result = parseCode(raw)
  }
  return result
}

const isMeitrack = raw => {
  let result = false
  if (patterns.mvt380.test(raw)) {
    result = true
  } else if (patterns.ok.test(raw)) {
    result = true
  }
  return result
}

// Random integer from 65 to 122 (41 to 7a in hex)
const getRandomDataIdentifier = () => {
  const int = Math.floor(Math.random() * (122 - 65 + 1) + 65)
  return String.fromCharCode(int)
}

const getCommand = (imei, command) => {
  const raw1 = `,${imei},${command}*`
  const raw2 = `@@${getRandomDataIdentifier()}${raw1.length + 4}${raw1}`
  return `${raw2}${pad(
    2,
    crc
      .crc1(raw2)
      .toString(16)
      .toUpperCase(),
    '0'
  )}\r\n`
}

const parseCommand = data => {
  let raw, state, port, speed, interval
  if (/^[1-5]{1}_(on|off|status)$/.test(data.instruction)) {
    let _data = data.instruction.split('_')
    port = _data[0]
    state = _data[1]
    let initial = [2, 2, 2, 2, 2]
    const states = { off: 0, on: 1, status: 2 }
    initial[port - 1] = states[state]
    speed = data.speed || 0
    raw = `C01,${speed},${initial.join('')}`
  } else if (data.instruction === 'clear_mem') {
    raw = 'F09,3'
  } else if (data.instruction === 'set_interval_gprs') {
    interval = data.interval || 6 * 10
    if (interval < 12) interval = 12
    let mod = interval % 6
    if (mod > 0) interval -= mod
    raw = `A12,${interval}`
  } else if (/^set_speed_(on|off)(E)?$/.test(data.instruction)) {
    speed = data.speed || 0
    state = data.instruction.split('_')[2]
    if (state === 'off') speed = 0
    raw = `B07,${speed}`
  } else if (data.instruction === 'Custom') {
    raw = data.command
  } else if (/^reboot_gsm$/.test(data.instruction)) {
    raw = 'F01'
  } else if (/^reboot_gps$/.test(data.instruction)) {
    raw = 'F02'
  }
  return getCommand(data.imei, raw)
}

const getRebootCommand = imei => {
  return getCommand(imei, 'F02')
}

const getImei = raw => {
  let imei = null
  const data = raw.toString()
  if (patterns.mvt380.test(data)) {
    imei = patterns.mvt380.exec(data)[3]
  }
  return imei
}

module.exports = {
  parse: parse,
  patterns: patterns,
  getMvt380: getMvt380,
  isMeitrack: isMeitrack,
  parseCommand: parseCommand,
  getRebootCommand: getRebootCommand,
  getImei: getImei
}
